File: python/src/livekit/plugins/ai_coustics/processor.py
from __future__ import annotations

import os
import time
from dataclasses import dataclass

import aic_sdk
import numpy as np

from livekit import rtc

from .log import logger

_SLOW_WARNING_INTERVAL = 10.0
_SLOW_BACKLOG_THRESHOLD = 0.2
_ERROR_REPORT_INTERVAL = 10.0


def _license_key(value: str | None) -> str:
    key = value or os.getenv("AIC_SDK_LICENSE")
    if not key:
        raise ValueError(
            "An ai-coustics SDK license is required. Pass license_key or set AIC_SDK_LICENSE."
        )
    return key


def _pcm16_to_float32(data: memoryview) -> np.ndarray:
    return np.frombuffer(data, dtype=np.int16).astype(np.float32) / 32768.0


def _float32_to_pcm16(data: np.ndarray) -> bytes:
    clipped = np.clip(data, -1.0, 32767.0 / 32768.0)
    return bytes(np.rint(clipped * 32768.0).astype(np.int16).tobytes())


@dataclass
class ProcessorParameters:
    """Runtime-adjustable enhancement parameters; ``None`` retains the current value."""

    enhancement_level: float | None = None
    bypass: bool | None = None


class Processor(rtc.FrameProcessor[rtc.AudioFrame]):
    """LiveKit audio frame processor backed by :class:`aic_sdk.Processor`.

    Model resolution and Processor construction are eager, while SDK backend authentication uses
    its grace period. Format initialization remains lazy because LiveKit supplies the stream's
    sample rate, channel count, and frame size with the first frame.
    """

    def __init__(
        self,
        *,
        model: aic_sdk.Model,
        license_key: str | None = None,
        processor_parameters: ProcessorParameters | None = None,
    ) -> None:
        resolved_license_key = _license_key(license_key)

        # This runtime-only SDK hook must run before Processor construction because the SDK keeps
        # the first integration identifier it receives. ID 8 identifies the LiveKit Python plugin.
        aic_sdk.set_sdk_id(8)  # type: ignore[attr-defined]

        try:
            model_id = model.get_id()
            processor = aic_sdk.Processor(
                model,
                resolved_license_key,
            )
        except Exception as error:
            raise RuntimeError(f"Failed to create ai-coustics Processor: {error}") from error

        self._processor: aic_sdk.Processor | None = processor
        self._model_id = model_id
        self._format: tuple[int, int, int] | None = None
        self._stream_info: dict[str, str] = {}
        self._enabled = True
        self._closed = False

        self._frame_count = 0
        self._processed_frame_count = 0
        self._failed_frame_count = 0
        self._input_audio_duration = 0.0
        self._processing_duration_total = 0.0
        self._processing_duration_max = 0.0
        self._sdk_processing_duration_total = 0.0
        self._sdk_processing_duration_max = 0.0
        self._realtime_factor_max = 0.0
        self._processing_backlog = 0.0
        self._processing_backlog_max = 0.0
        self._initialization_count = 0
        self._audio_delay_samples: int | None = None
        self._audio_delay: float | None = None
        self._last_slow_warning: float | None = None
        self._slow_warning_active = False

        self._consecutive_failures = 0
        self._failure_started: float | None = None
        self._failure_episode_reported = False
        self._active_error_signature: tuple[str, str, str] | None = None
        self._last_reported_error_signature: tuple[str, str, str] | None = None
        self._last_error_report: float | None = None

        try:
            self._context: aic_sdk.ProcessorContext | None = processor.get_context()
            if processor_parameters is not None:
                self.set_parameters(processor_parameters)
        except Exception:
            processor.terminate_session()
            raise

    @property
    def enabled(self) -> bool:
        return self._enabled

    @enabled.setter
    def enabled(self, value: bool) -> None:
        if value == self._enabled:
            return
        if value and not self._enabled and self._context is not None:
            self._context.reset()
        self._enabled = value
        logger.debug(
            "ai-coustics Processor %s",
            "enabled" if value else "disabled",
            extra=self._diagnostic_fields(),
        )

    def set_parameters(self, parameters: ProcessorParameters) -> None:
        """Apply a partial Processor-parameter update."""

        if self._context is None:
            return

        if parameters.enhancement_level is not None:
            level = parameters.enhancement_level
            if not 0.0 <= level <= 1.0:
                raise ValueError(f"enhancement_level must be in [0.0, 1.0], got {level}")
            self._context.set_parameter(aic_sdk.ProcessorParameter.EnhancementLevel, level)
            logger.debug(
                "ai-coustics Processor parameter updated",
                extra=self._diagnostic_fields(parameter="enhancement_level", parameter_value=level),
            )

        if parameters.bypass is not None:
            if not isinstance(parameters.bypass, bool):
                raise TypeError("bypass must be a bool")
            self._context.set_parameter(
                aic_sdk.ProcessorParameter.Bypass,
                1.0 if parameters.bypass else 0.0,
            )
            logger.debug(
                "ai-coustics Processor parameter updated",
                extra=self._diagnostic_fields(
                    parameter="bypass", parameter_value=parameters.bypass
                ),
            )

    def _on_stream_info_updated(
        self,
        *,
        room_name: str,
        participant_identity: str,
        publication_sid: str,
    ) -> None:
        stream_info = {
            "room_name": room_name,
            "participant_identity": participant_identity,
            "publication_sid": publication_sid,
        }
        changed = bool(self._stream_info and self._stream_info != stream_info)
        if changed and self._context is not None:
            self._context.reset()
        self._stream_info = stream_info
        logger.debug(
            "ai-coustics Processor stream %s",
            "changed; native context reset" if changed else "attached",
            extra=self._diagnostic_fields(),
        )

    def _on_stream_info_cleared(self) -> None:
        if not self._stream_info:
            return
        fields = self._diagnostic_fields()
        if self._context is not None:
            self._context.reset()
        self._stream_info = {}
        logger.debug("ai-coustics Processor stream detached", extra=fields)

    def _process(self, frame: rtc.AudioFrame) -> rtc.AudioFrame:
        # A disabled or closed processor is a transparent LiveKit frame processor.
        if not self.enabled or self._processor is None or self._context is None:
            return frame

        started = time.perf_counter()
        frame_duration = frame.samples_per_channel / frame.sample_rate
        processing_stage = "initialize"
        sdk_processing_duration = 0.0
        try:
            # LiveKit supplies the stream geometry with each frame, so SDK initialization must be
            # lazy. Reinitializing on a geometry change also clears state from the previous stream.
            stream_format = (
                frame.sample_rate,
                frame.num_channels,
                frame.samples_per_channel,
            )

            if self._format != stream_format:
                initialization_started = time.perf_counter()
                self._processor.initialize(
                    aic_sdk.ProcessorConfig(
                        sample_rate=frame.sample_rate,
                        block_size=frame.samples_per_channel,
                        variable_block_size=False,
                    )
                )
                self._format = stream_format
                self._initialization_count += 1
                initialization_duration = time.perf_counter() - initialization_started
                audio_delay_samples = self._context.get_audio_delay()
                self._audio_delay_samples = audio_delay_samples
                self._audio_delay = audio_delay_samples / frame.sample_rate
                logger.info(
                    "ai-coustics Processor %s",
                    "initialized" if self._initialization_count == 1 else "reconfigured",
                    extra=self._diagnostic_fields(
                        initialization_duration=initialization_duration,
                        initialization_count=self._initialization_count,
                        audio_delay_samples=audio_delay_samples,
                        audio_delay=self._audio_delay,
                        frame_duration=frame_duration,
                        downmixing=frame.num_channels > 1,
                    ),
                )

            # LiveKit stores signed 16-bit PCM samples interleaved by channel. The SDK operates on
            # normalized float32 samples, so validate the frame before reshaping it.
            processing_stage = "validate_frame"
            samples = _pcm16_to_float32(frame.data)
            expected_samples = frame.samples_per_channel * frame.num_channels
            if samples.size != expected_samples:
                raise ValueError(
                    f"AudioFrame contains {samples.size} samples, expected {expected_samples}"
                )

            # aic-sdk 3 accepts only a contiguous, one-dimensional mono block. For multichannel
            # input, average all channels just as the pre-3.0 SDK did internally. Copying the mono
            # view makes it contiguous as well.
            channels = samples.reshape(frame.samples_per_channel, frame.num_channels)
            mono = (
                channels[:, 0].copy()
                if frame.num_channels == 1
                else channels.mean(axis=1, dtype=np.float32)
            )
            processing_stage = "process"
            sdk_started = time.perf_counter()
            try:
                processed = self._processor.process(mono)
            finally:
                sdk_processing_duration = time.perf_counter() - sdk_started

            # FrameProcessor output must retain the input geometry. Expand the enhanced mono block
            # back to interleaved PCM by writing the same processed signal to every input channel.
            processing_stage = "convert_output"
            interleaved = (
                processed if frame.num_channels == 1 else np.repeat(processed, frame.num_channels)
            )
            output = rtc.AudioFrame(
                data=_float32_to_pcm16(interleaved),
                sample_rate=frame.sample_rate,
                num_channels=frame.num_channels,
                samples_per_channel=frame.samples_per_channel,
                userdata=frame.userdata,
            )
        except Exception as error:
            # Keep room audio flowing if SDK initialization or processing fails.
            completed = time.perf_counter()
            self._record_timing(
                processing_duration=completed - started,
                sdk_processing_duration=sdk_processing_duration,
                frame_duration=frame_duration,
                completed=completed,
            )
            self._record_failure(
                processing_stage,
                error,
                completed,
                processing_duration=completed - started,
                sdk_processing_duration=sdk_processing_duration,
                frame_duration=frame_duration,
            )
            return frame

        completed = time.perf_counter()
        self._record_timing(
            processing_duration=completed - started,
            sdk_processing_duration=sdk_processing_duration,
            frame_duration=frame_duration,
            completed=completed,
        )
        self._record_success(completed)
        return output

    def _record_timing(
        self,
        *,
        processing_duration: float,
        sdk_processing_duration: float,
        frame_duration: float,
        completed: float,
    ) -> None:
        self._frame_count += 1
        self._input_audio_duration += frame_duration
        self._processing_duration_total += processing_duration
        self._processing_duration_max = max(self._processing_duration_max, processing_duration)
        self._sdk_processing_duration_total += sdk_processing_duration
        self._sdk_processing_duration_max = max(
            self._sdk_processing_duration_max, sdk_processing_duration
        )
        realtime_factor = processing_duration / frame_duration
        self._realtime_factor_max = max(self._realtime_factor_max, realtime_factor)
        self._processing_backlog = max(
            0.0,
            self._processing_backlog + processing_duration - frame_duration,
        )
        self._processing_backlog_max = max(self._processing_backlog_max, self._processing_backlog)

        if self._processing_backlog == 0.0:
            self._slow_warning_active = False
        elif self._processing_backlog >= _SLOW_BACKLOG_THRESHOLD and (
            not self._slow_warning_active
            or self._last_slow_warning is None
            or completed - self._last_slow_warning >= _SLOW_WARNING_INTERVAL
        ):
            self._slow_warning_active = True
            self._last_slow_warning = completed
            logger.warning(
                "ai-coustics Processor is falling behind realtime",
                extra=self._diagnostic_fields(
                    processing_duration=processing_duration,
                    sdk_processing_duration=sdk_processing_duration,
                    frame_duration=frame_duration,
                    realtime_factor=realtime_factor,
                    processing_backlog=self._processing_backlog,
                    processing_backlog_max=self._processing_backlog_max,
                ),
            )

    def _record_failure(
        self,
        stage: str,
        error: Exception,
        completed: float,
        *,
        processing_duration: float,
        sdk_processing_duration: float,
        frame_duration: float,
    ) -> None:
        signature = (stage, type(error).__name__, str(error))
        self._failed_frame_count += 1
        if self._consecutive_failures == 0:
            self._failure_started = completed - processing_duration
            self._failure_episode_reported = False
        self._consecutive_failures += 1
        self._active_error_signature = signature

        should_report = (
            signature != self._last_reported_error_signature
            or self._last_error_report is None
            or completed - self._last_error_report >= _ERROR_REPORT_INTERVAL
        )
        if not should_report:
            return

        self._failure_episode_reported = True
        self._last_reported_error_signature = signature
        self._last_error_report = completed
        logger.error(
            "ai-coustics Processor failed; passing audio through",
            extra=self._diagnostic_fields(
                processing_stage=stage,
                error_type=type(error).__name__,
                error_message=str(error),
                consecutive_failures=self._consecutive_failures,
                failed_frame_count=self._failed_frame_count,
                processing_duration=processing_duration,
                sdk_processing_duration=sdk_processing_duration,
                frame_duration=frame_duration,
                realtime_factor=processing_duration / frame_duration,
                failure_duration=completed
                - (self._failure_started if self._failure_started is not None else completed),
            ),
            exc_info=(type(error), error, error.__traceback__),
        )

    def _record_success(self, completed: float) -> None:
        self._processed_frame_count += 1
        if self._consecutive_failures == 0:
            return

        if self._failure_episode_reported:
            active_error = self._active_error_signature
            logger.info(
                "ai-coustics Processor recovered",
                extra=self._diagnostic_fields(
                    recovered_failure_count=self._consecutive_failures,
                    failure_duration=completed
                    - (self._failure_started if self._failure_started is not None else completed),
                    failed_frame_count=self._failed_frame_count,
                    last_failure_stage=active_error[0] if active_error else None,
                    last_error_type=active_error[1] if active_error else None,
                    last_error_message=active_error[2] if active_error else None,
                ),
            )
        self._consecutive_failures = 0
        self._failure_started = None
        self._failure_episode_reported = False
        self._active_error_signature = None

    def _diagnostic_fields(self, **fields: object) -> dict[str, object]:
        diagnostics: dict[str, object] = {
            "model_provider": "ai-coustics",
            "model_name": self._model_id,
            **self._stream_info,
        }
        if self._format is not None:
            diagnostics.update(
                {
                    "sample_rate": self._format[0],
                    "num_channels": self._format[1],
                    "samples_per_frame": self._format[2],
                }
            )
        diagnostics.update(fields)
        return diagnostics

    def _close(self) -> None:
        if self._closed:
            return
        self._closed = True
        self._enabled = False
        processor = self._processor

        if processor is not None:
            try:
                processor.terminate_session()
            except Exception as error:
                logger.error(
                    "Failed to terminate ai-coustics Processor session",
                    extra=self._diagnostic_fields(
                        error_type=type(error).__name__, error_message=str(error)
                    ),
                    exc_info=(type(error), error, error.__traceback__),
                )

        summary = self._diagnostic_fields(
            frame_count=self._frame_count,
            processed_frame_count=self._processed_frame_count,
            failed_frame_count=self._failed_frame_count,
            input_audio_duration=self._input_audio_duration,
            processing_duration_total=self._processing_duration_total,
            processing_duration_max=self._processing_duration_max,
            sdk_processing_duration_total=self._sdk_processing_duration_total,
            sdk_processing_duration_max=self._sdk_processing_duration_max,
            average_realtime_factor=(
                self._processing_duration_total / self._input_audio_duration
                if self._input_audio_duration > 0.0
                else 0.0
            ),
            maximum_realtime_factor=self._realtime_factor_max,
            processing_backlog_max=self._processing_backlog_max,
            initialization_count=self._initialization_count,
            audio_delay_samples=self._audio_delay_samples,
            audio_delay=self._audio_delay,
            consecutive_failures=self._consecutive_failures,
            active_failure_stage=(
                self._active_error_signature[0]
                if self._active_error_signature is not None
                else None
            ),
            active_error_type=(
                self._active_error_signature[1]
                if self._active_error_signature is not None
                else None
            ),
            active_error_message=(
                self._active_error_signature[2]
                if self._active_error_signature is not None
                else None
            ),
        )
        if self._frame_count:
            logger.info("ai-coustics Processor closed", extra=summary)
        else:
            logger.debug("ai-coustics Processor closed without processing audio", extra=summary)

        self._context = None
        self._processor = None
        self._format = None

File: node/src/processor.ts
import { log } from "@livekit/agents";
import {
  AudioFrame,
  FrameProcessor,
  type FrameProcessorStreamInfo,
} from "@livekit/rtc-node";

import {
  type Model,
  Processor as AicProcessor,
  type ProcessorContext,
  ProcessorParameter as AicProcessorParameter,
  setSdkId,
} from "./sdk.js";

const SLOW_WARNING_INTERVAL_MS = 10_000;
const SLOW_BACKLOG_THRESHOLD_MS = 200;
const ERROR_REPORT_INTERVAL_MS = 10_000;

type ProcessingStage = "initialize" | "validate_frame" | "process" | "convert_output";
type LogLevel = "debug" | "info" | "warn" | "error";

export interface ProcessorParameters {
  enhancementLevel?: number;
  bypass?: boolean;
}

export interface ProcessorOptions {
  /** Loaded ai-coustics SDK Model. */
  model: Model;
  licenseKey?: string;
  processorParameters?: ProcessorParameters;
}

function resolveLicenseKey(value?: string): string {
  const key = value || process.env.AIC_SDK_LICENSE;
  if (!key) {
    throw new Error(
      "An ai-coustics SDK license is required. Pass licenseKey or set AIC_SDK_LICENSE.",
    );
  }
  return key;
}

export function pcm16ToFloat32(data: Int16Array): Float32Array {
  return Float32Array.from(data, (sample) => sample / 32768);
}

export function float32ToPcm16(data: Float32Array): Int16Array {
  return Int16Array.from(data, (sample) => {
    const clipped = Math.max(-1, Math.min(32767 / 32768, sample));
    return Math.round(clipped * 32768);
  });
}

/** LiveKit AudioFrame processor backed by the public ai-coustics SDK. */
export class Processor extends FrameProcessor<AudioFrame> {
  private processor: InstanceType<typeof AicProcessor> | null;
  private context: ProcessorContext | null;
  private readonly modelId: string;
  private streamFormat: [number, number, number] | null = null;
  private streamInfo: FrameProcessorStreamInfo | null = null;
  private filteringEnabled = true;

  private frameCount = 0;
  private processedFrameCount = 0;
  private failedFrameCount = 0;
  private inputAudioDurationMs = 0;
  private processingDurationTotalMs = 0;
  private processingDurationMaxMs = 0;
  private sdkProcessingDurationTotalMs = 0;
  private sdkProcessingDurationMaxMs = 0;
  private maximumRealtimeFactor = 0;
  private processingBacklogMs = 0;
  private processingBacklogMaxMs = 0;
  private initializationCount = 0;
  private audioDelaySamples: number | null = null;
  private audioDelayMs: number | null = null;
  private lastSlowWarning: number | null = null;
  private slowWarningActive = false;

  private consecutiveFailures = 0;
  private failureStarted: number | null = null;
  private failureEpisodeReported = false;
  private activeErrorSignature: string | null = null;
  private lastReportedErrorSignature: string | null = null;
  private lastErrorReport: number | null = null;

  constructor(options: ProcessorOptions) {
    super();
    const licenseKey = resolveLicenseKey(options.licenseKey);

    // The SDK keeps the first integration identifier it receives. Set this before Processor
    // construction so usage is attributed to the LiveKit Node plugin.
    setSdkId(9);

    try {
      this.modelId = options.model.getId();
      this.processor = new AicProcessor(options.model, licenseKey);
    } catch (error) {
      const detail = error instanceof Error ? error.message : String(error);
      throw new Error(`Failed to create ai-coustics Processor: ${detail}`, {
        cause: error,
      });
    }
    try {
      this.context = this.processor.getContext();
      if (options.processorParameters) {
        this.setParameters(options.processorParameters);
      }
    } catch (error) {
      try {
        this.processor.terminateSession();
      } catch {
        // best-effort release of the freshly-allocated native session
      }
      throw error;
    }
  }

  isEnabled(): boolean {
    return this.filteringEnabled;
  }

  setEnabled(enabled: boolean): void {
    if (enabled === this.filteringEnabled) return;
    if (enabled && !this.filteringEnabled && this.context) {
      this.context.reset();
    }
    this.filteringEnabled = enabled;
    this.writeLog(
      "debug",
      `ai-coustics Processor ${enabled ? "enabled" : "disabled"}`,
    );
  }

  setParameters(parameters: ProcessorParameters): void {
    if (!this.context) return;
    if (parameters.enhancementLevel !== undefined) {
      const level = parameters.enhancementLevel;
      if (level < 0 || level > 1) {
        throw new Error(`enhancementLevel must be in [0.0, 1.0], got ${level}`);
      }
      this.context.setParameter(AicProcessorParameter.EnhancementLevel, level);
      this.writeLog("debug", "ai-coustics Processor parameter updated", {
        parameter: "enhancementLevel",
        parameterValue: level,
      });
    }
    if (parameters.bypass !== undefined) {
      if (typeof parameters.bypass !== "boolean") {
        throw new TypeError("bypass must be a boolean");
      }
      this.context.setParameter(
        AicProcessorParameter.Bypass,
        parameters.bypass ? 1 : 0,
      );
      this.writeLog("debug", "ai-coustics Processor parameter updated", {
        parameter: "bypass",
        parameterValue: parameters.bypass,
      });
    }
  }

  override onStreamInfoUpdated(info: FrameProcessorStreamInfo): void {
    const changed =
      this.streamInfo !== null &&
      (this.streamInfo.roomName !== info.roomName ||
        this.streamInfo.participantIdentity !== info.participantIdentity ||
        this.streamInfo.publicationSid !== info.publicationSid);
    if (changed) this.context?.reset();
    this.streamInfo = { ...info };
    this.writeLog(
      "debug",
      `ai-coustics Processor stream ${changed ? "changed; native context reset" : "attached"}`,
    );
  }

  override onStreamInfoCleared(): void {
    if (!this.streamInfo) return;
    const fields = this.diagnosticFields();
    this.context?.reset();
    this.streamInfo = null;
    this.writeLog("debug", "ai-coustics Processor stream detached", fields, true);
  }

  process(frame: AudioFrame): AudioFrame {
    if (!this.isEnabled() || !this.processor || !this.context) {
      return frame;
    }

    const started = performance.now();
    const frameDurationMs =
      (frame.samplesPerChannel / frame.sampleRate) * 1000;
    let processingStage: ProcessingStage = "initialize";
    let sdkProcessingDurationMs = 0;
    try {
      const streamFormat: [number, number, number] = [
        frame.sampleRate,
        frame.channels,
        frame.samplesPerChannel,
      ];
      if (
        !this.streamFormat ||
        this.streamFormat[0] !== streamFormat[0] ||
        this.streamFormat[1] !== streamFormat[1] ||
        this.streamFormat[2] !== streamFormat[2]
      ) {
        const initializationStarted = performance.now();
        this.processor.initialize(streamFormat[0], streamFormat[2], false);
        this.streamFormat = streamFormat;
        this.initializationCount += 1;
        const initializationDurationMs = performance.now() - initializationStarted;
        const audioDelaySamples = this.context.getAudioDelay();
        this.audioDelaySamples = audioDelaySamples;
        this.audioDelayMs = (audioDelaySamples / frame.sampleRate) * 1000;
        this.writeLog(
          "info",
          `ai-coustics Processor ${this.initializationCount === 1 ? "initialized" : "reconfigured"}`,
          {
            initializationDurationMs,
            initializationCount: this.initializationCount,
            audioDelaySamples,
            audioDelayMs: this.audioDelayMs,
            frameDurationMs,
            downmixing: frame.channels > 1,
          },
        );
      }
      processingStage = "validate_frame";
      const expectedSamples = frame.samplesPerChannel * frame.channels;
      if (frame.data.length !== expectedSamples) {
        throw new Error(
          `AudioFrame contains ${frame.data.length} samples, expected ${expectedSamples}`,
        );
      }

      const samples = pcm16ToFloat32(frame.data);
      const mono = new Float32Array(frame.samplesPerChannel);
      if (frame.channels === 1) {
        mono.set(samples);
      } else {
        for (let sample = 0; sample < frame.samplesPerChannel; sample += 1) {
          let sum = 0;
          for (let channel = 0; channel < frame.channels; channel += 1) {
            sum += samples[sample * frame.channels + channel]!;
          }
          mono[sample] = sum / frame.channels;
        }
      }

      processingStage = "process";
      const sdkStarted = performance.now();
      try {
        this.processor.process(mono);
      } finally {
        sdkProcessingDurationMs = performance.now() - sdkStarted;
      }
      processingStage = "convert_output";
      const processedMono = float32ToPcm16(mono);
      let data: Int16Array;
      if (frame.channels === 1) {
        data = processedMono;
      } else {
        data = new Int16Array(expectedSamples);
        for (let sample = 0; sample < frame.samplesPerChannel; sample += 1) {
          for (let channel = 0; channel < frame.channels; channel += 1) {
            data[sample * frame.channels + channel] = processedMono[sample]!;
          }
        }
      }
      const output = new AudioFrame(
        data,
        frame.sampleRate,
        frame.channels,
        frame.samplesPerChannel,
        frame.userdata,
      );
      const completed = performance.now();
      this.recordTiming(
        completed - started,
        sdkProcessingDurationMs,
        frameDurationMs,
        completed,
      );
      this.recordSuccess(completed);
      return output;
    } catch (error) {
      const completed = performance.now();
      this.recordTiming(
        completed - started,
        sdkProcessingDurationMs,
        frameDurationMs,
        completed,
      );
      this.recordFailure(
        processingStage,
        error,
        completed,
        completed - started,
        sdkProcessingDurationMs,
        frameDurationMs,
      );
      return frame;
    }
  }

  close(): void {
    if (!this.processor) return;
    this.filteringEnabled = false;
    const processor = this.processor;

    try {
      processor.terminateSession();
    } catch (error) {
      this.writeLog(
        "error",
        "Failed to terminate ai-coustics Processor session",
        {},
        false,
        error,
      );
    }

    const summary = this.diagnosticFields({
      frameCount: this.frameCount,
      processedFrameCount: this.processedFrameCount,
      failedFrameCount: this.failedFrameCount,
      inputAudioDurationMs: this.inputAudioDurationMs,
      processingDurationTotalMs: this.processingDurationTotalMs,
      processingDurationMaxMs: this.processingDurationMaxMs,
      sdkProcessingDurationTotalMs: this.sdkProcessingDurationTotalMs,
      sdkProcessingDurationMaxMs: this.sdkProcessingDurationMaxMs,
      averageRealtimeFactor:
        this.inputAudioDurationMs > 0
          ? this.processingDurationTotalMs / this.inputAudioDurationMs
          : 0,
      maximumRealtimeFactor: this.maximumRealtimeFactor,
      processingBacklogMaxMs: this.processingBacklogMaxMs,
      initializationCount: this.initializationCount,
      audioDelaySamples: this.audioDelaySamples,
      audioDelayMs: this.audioDelayMs,
      consecutiveFailures: this.consecutiveFailures,
      activeFailureStage: this.activeErrorSignature?.split("\u0000")[0],
      activeErrorType: this.activeErrorSignature?.split("\u0000")[1],
      activeErrorMessage: this.activeErrorSignature?.split("\u0000")[2],
    });
    this.writeLog(
      this.frameCount > 0 ? "info" : "debug",
      this.frameCount > 0
        ? "ai-coustics Processor closed"
        : "ai-coustics Processor closed without processing audio",
      summary,
      true,
    );
    this.context = null;
    this.processor = null;
    this.streamFormat = null;
    this.streamInfo = null;
  }

  private recordTiming(
    processingDurationMs: number,
    sdkProcessingDurationMs: number,
    frameDurationMs: number,
    completed: number,
  ): void {
    this.frameCount += 1;
    this.inputAudioDurationMs += frameDurationMs;
    this.processingDurationTotalMs += processingDurationMs;
    this.processingDurationMaxMs = Math.max(
      this.processingDurationMaxMs,
      processingDurationMs,
    );
    this.sdkProcessingDurationTotalMs += sdkProcessingDurationMs;
    this.sdkProcessingDurationMaxMs = Math.max(
      this.sdkProcessingDurationMaxMs,
      sdkProcessingDurationMs,
    );
    const realtimeFactor = processingDurationMs / frameDurationMs;
    this.maximumRealtimeFactor = Math.max(
      this.maximumRealtimeFactor,
      realtimeFactor,
    );
    this.processingBacklogMs = Math.max(
      0,
      this.processingBacklogMs + processingDurationMs - frameDurationMs,
    );
    this.processingBacklogMaxMs = Math.max(
      this.processingBacklogMaxMs,
      this.processingBacklogMs,
    );

    if (this.processingBacklogMs === 0) {
      this.slowWarningActive = false;
    } else if (
      this.processingBacklogMs >= SLOW_BACKLOG_THRESHOLD_MS &&
      (!this.slowWarningActive ||
        this.lastSlowWarning === null ||
        completed - this.lastSlowWarning >= SLOW_WARNING_INTERVAL_MS)
    ) {
      this.slowWarningActive = true;
      this.lastSlowWarning = completed;
      this.writeLog("warn", "ai-coustics Processor is falling behind realtime", {
        processingDurationMs,
        sdkProcessingDurationMs,
        frameDurationMs,
        realtimeFactor,
        processingBacklogMs: this.processingBacklogMs,
        processingBacklogMaxMs: this.processingBacklogMaxMs,
      });
    }
  }

  private recordFailure(
    stage: ProcessingStage,
    error: unknown,
    completed: number,
    processingDurationMs: number,
    sdkProcessingDurationMs: number,
    frameDurationMs: number,
  ): void {
    const errorType = error instanceof Error ? error.name : typeof error;
    const errorMessage = error instanceof Error ? error.message : String(error);
    const signature = `${stage}\u0000${errorType}\u0000${errorMessage}`;
    this.failedFrameCount += 1;
    if (this.consecutiveFailures === 0) {
      this.failureStarted = completed - processingDurationMs;
      this.failureEpisodeReported = false;
    }
    this.consecutiveFailures += 1;
    this.activeErrorSignature = signature;

    const shouldReport =
      signature !== this.lastReportedErrorSignature ||
      this.lastErrorReport === null ||
      completed - this.lastErrorReport >= ERROR_REPORT_INTERVAL_MS;
    if (!shouldReport) return;

    this.failureEpisodeReported = true;
    this.lastReportedErrorSignature = signature;
    this.lastErrorReport = completed;
    this.writeLog(
      "error",
      "ai-coustics Processor failed; passing audio through",
      {
        processingStage: stage,
        errorType,
        errorMessage,
        consecutiveFailures: this.consecutiveFailures,
        failedFrameCount: this.failedFrameCount,
        processingDurationMs,
        sdkProcessingDurationMs,
        frameDurationMs,
        realtimeFactor: processingDurationMs / frameDurationMs,
        failureDurationMs:
          completed - (this.failureStarted === null ? completed : this.failureStarted),
      },
      false,
      error,
    );
  }

  private recordSuccess(completed: number): void {
    this.processedFrameCount += 1;
    if (this.consecutiveFailures === 0) return;

    if (this.failureEpisodeReported) {
      const [lastFailureStage, lastErrorType, lastErrorMessage] =
        this.activeErrorSignature?.split("\u0000") ?? [];
      this.writeLog("info", "ai-coustics Processor recovered", {
        recoveredFailureCount: this.consecutiveFailures,
        failureDurationMs:
          completed - (this.failureStarted === null ? completed : this.failureStarted),
        failedFrameCount: this.failedFrameCount,
        lastFailureStage,
        lastErrorType,
        lastErrorMessage,
      });
    }
    this.consecutiveFailures = 0;
    this.failureStarted = null;
    this.failureEpisodeReported = false;
    this.activeErrorSignature = null;
  }

  private diagnosticFields(
    fields: Record<string, unknown> = {},
  ): Record<string, unknown> {
    const diagnostics: Record<string, unknown> = {
      modelProvider: "ai-coustics",
      modelName: this.modelId,
    };
    if (this.streamInfo) Object.assign(diagnostics, this.streamInfo);
    if (this.streamFormat) {
      Object.assign(diagnostics, {
        sampleRate: this.streamFormat[0],
        numChannels: this.streamFormat[1],
        samplesPerFrame: this.streamFormat[2],
      });
    }
    return Object.assign(diagnostics, fields);
  }

  private writeLog(
    level: LogLevel,
    message: string,
    fields: Record<string, unknown> = {},
    fieldsAreComplete = false,
    error?: unknown,
  ): void {
    const diagnostics = fieldsAreComplete
      ? fields
      : this.diagnosticFields(fields);
    const payload = error === undefined ? diagnostics : { ...diagnostics, error };
    try {
      log()[level](payload, message);
      return;
    } catch (loggingError) {
      const loggerIsUninitialized =
        loggingError instanceof TypeError &&
        loggingError.message.includes("logger not initialized");
      if (!loggerIsUninitialized) {
        console.error("Failed to write ai-coustics diagnostic through LiveKit", {
          error: loggingError,
        });
      }
    }

    // Processor instances can be constructed before LiveKit configures its global logger. Keep
    // operational diagnostics visible in that case, but do not turn normally-hidden debug events
    // into unsolicited console output.
    if (level === "error") console.error(message, payload);
    else if (level === "warn") console.warn(message, payload);
    else if (level === "info") console.info(message, payload);
  }
}
